external-crates/move/crates/move-package-alt/DESIGN.md

```diff
@@ -295,61 +295,119 @@ AddressInfo:
 # Internal Operations
 
-## Overview
+## Pinned dependencies
 
-The purpose of the lockfile is to give stability to developers by pinning consistent dependency
-versions. These pinned versions are then used by other developers working on the same project, by
-CI, and so forth. Dependencies are only repinned in two situations:
+The purpose of the lockfile is to give stability to developers by pinning consistent (transitive)
+dependency versions. These pinned versions are then used by other developers working on the same
+project, by CI, by source verifiers, and so forth.
+
+A pinned dependency is one that guarantees that future lookups of the same dependency
+will yield the same source package. For example, a git dependency with a branch revision is not
+pinned, because the branch may be moved to a different commit, which might have different source
+code. To pin this kind of dependency, we must replace the branch name with a commit hash -
+downloading the same commit should always produce the same source code.
+
+Although we expect pinned dependencies to not change, there are a few situations where they could.
+One possiblity is local dependencies - if a single repository contains both a dependency and the
+depending project, we would want changes in the dependency to be reflected immediately in the
+depending project. Moreover, this doesn't cause inconsistencies in CI because presumably those
+changes would be committed together. Another example would be if a user wants to temporarily change
+a cached dependency during debugging to rerun a test or something. To handle these situations, we
+will store digests of all transitive dependency manifests and repin if any of them change.
+
+Dependencies are always pinned as a group and are only repinned in two situations:
 
 1. The user explicitly asks for it by running `sui move update-deps`. This command will repin all
-   dependencies (or you could ask to update a single dependency using `sui move update-deps
-   dep-name`).
-2. The user adds or modifies a dependency in their `Move.toml` file. We detect this by comparing
-   `unpublished.dependencies.unpinned` in `Move.lock` with `dependencies` in `Move.toml`. If an
-   entry is different, that means the user has modified it, so we repin it.
+   dependencies.
+
+2. The user modifies a `Move.toml` file in their dependency tree. We detect this by comparing
+   the `pinned.environment.ID.manifest_digest` field of the lockfile of the project being built to the manifest
+   file of the corresponding dependency.
+
+Note: we had considered only repinning the dependencies that had changed and allowing the user to
+repin only specific deps, but this leads to a lot of confusing corner cases. This is consistent with
+the philosophy is that pinning is a mechanism for reliable rebuilding and not for dependency version
+management - if a user cares about the exact version of a dependency then they can indicate that by
+referring to that version in their manifest.
```

Comment:
We've also discussed solving other problems by allowing multiple source packages for the same on-chain package; this would also simplify the transitive dependency repinning, which would allow us to do repinning package-by-package. Maybe we should actually keep the old design here given that solution